src/storage/redis_db.cc

```diff
@@ -359,28 +363,57 @@ rocksdb::Status Database::Scan(const std::string &cursor, uint64_t limit, const
   }
 
   uint16_t slot_id = slot_start;
+
+  auto scan_match_check = [&]() -> bool {
+    if (ns_prefix.empty()) {
+      return true;
+    }
+    if (!iter->key().starts_with(ns_prefix)) {
+      return false;
+    }
+    auto key_view = iter->key().ToStringView();
+    auto sub_key_view = static_cast<Slice>(key_view.substr(ns_prefix.size(), key_view.size() - ns_prefix.size()));
+    if (pm == 1 && !sub_key_view.ends_with(fix)) {
+      return false;
+    } else if (pm == 2 && sub_key_view.ToStringView().find(fix) == std::string::npos) {
+      return false;
+    }
+    return true;
+  };
+
   while (true) {
-    for (; iter->Valid() && cnt < limit; iter->Next()) {
-      if (!ns_prefix.empty() && !iter->key().starts_with(ns_prefix)) {
-        break;
+    for (; iter->Valid() && cnt_success < scan_success_limit; iter->Next()) {
+      if (!scan_match_check()) {
+        cnt_false++;
+        if (cnt_false < scan_false_limit) {
+          continue;
+        }
       }
+      // The user_key must be updated either when the scan is successful, or when the number of consecutive failures
+      // reaches its limit
       Metadata metadata(kRedisNone, false);
       auto s = metadata.Decode(iter->value());
       if (!s.ok()) continue;
 
       if (metadata.Expired()) continue;
       std::tie(std::ignore, user_key) = ExtractNamespaceKey<std::string>(iter->key(), storage_->IsSlotIdEncoded());
+      if (cnt_false >= scan_false_limit) {
+        break;
+      }
       keys->emplace_back(user_key);
-      cnt++;
+      cnt_success++;
     }
-    if (!storage_->IsSlotIdEncoded() || prefix.empty()) {
-      if (!keys->empty() && cnt >= limit) {
+    if (!storage_->IsSlotIdEncoded() || fix.empty()) {
+      if (!keys->empty() && (cnt_success >= scan_success_limit)) {
+        end_cursor->append(user_key);
+      }
+      if (keys->empty() && cnt_false >= scan_false_limit) {
         end_cursor->append(user_key);
       }
```

Comment:
There may be a situation where some keys have been found, but `cnt_false >= scan_false_limit`.